ribbon-eureka/src/main/java/com/netflix/loadbalancer/EurekaDynamicNotificationLoadBalancer.java

```diff
@@ -0,0 +1,159 @@
+package com.netflix.loadbalancer;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.netflix.client.config.IClientConfig;
+import com.netflix.discovery.CacheRefreshedEvent;
+import com.netflix.discovery.EurekaClient;
+import com.netflix.discovery.EurekaEvent;
+import com.netflix.discovery.EurekaEventListener;
+import com.netflix.niws.loadbalancer.DiscoveryEnabledNIWSServerList;
+import com.netflix.niws.loadbalancer.DiscoveryEnabledServer;
+import com.netflix.servo.annotations.DataSourceType;
+import com.netflix.servo.annotations.Monitor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Provider;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * A eureka specific dynamic loadbalancer that can update it's internal server list via notifications
+ * of cache updates from {@link com.netflix.discovery.EurekaClient}.
+ *
+ * Note that when a cache refreshed notification is received, the actual update on the serverList is
+ * done on a separate scheduler as the notification is delivered on an eurekaClient thread.
+ *
+ * @author David Liu
+ */
+public class EurekaDynamicNotificationLoadBalancer extends AbstractDynamicServerListLoadBalancer<DiscoveryEnabledServer> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EurekaDynamicNotificationLoadBalancer.class);
+
+    public static final ExecutorService DEFAULT_SERVER_LIST_UPDATE_EXECUTOR = Executors.newCachedThreadPool(
```

Comment:
Make this lazy https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom